libvast/vast/detail/shutdown_stream_stage.hpp

```diff
@@ -1,26 +1,49 @@
-#pragma once
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
 
-#include <caf/stream_stage.hpp>
+#pragma once
 
-#include <type_traits>
+#include <vast/concept/support/detail/is_specialization_of.hpp>
 
-template <template <class...> class Template, class Instantiation>
-struct is_specialization_of : std::false_type {};
+#include <caf/stream_stage.hpp>
 
-template <template <class...> class Template, class... Args>
-struct is_specialization_of<Template, Template<Args...>> : std::true_type {};
+namespace vast::detail {
 
-template <typename StreamStage>
-void shutdown_stream_stage(StreamStage& stage) requires(
-  is_specialization_of<caf::stream_stage, StreamStage>::value) {
-  // First we call `shutdown()` to notify all upstream
+/// Properly flushes and shuts down a `caf::stream_stage` connected to
+/// a `caf::broadcast_downstream_manager`.
+template <typename In, typename DownstreamManager>
+void shutdown_stream_stage(
+  caf::stream_stage_ptr<In, DownstreamManager>&
+    stage) requires(is_specialization_of<caf::broadcast_downstream_manager,
+                                         DownstreamManager>::value) {
+  if (!stage)
+    return;
```

Comment:
Should this function instead take the stream stage by reference?

Reply:
I'm not sure, it's always stored as a `stream_stage_ptr` so it seems more natural to just pass that instead of letting every call site check for null.

Reply:
Just thinking about the contract of the function.

```cpp
// As a developer I know this is a no-op if the stage is a nullptr.
shutdown_stream_stage(self->state.stage);
```

vs.

```cpp
// The precondition that the stage must still be valid is immediately
// visible at the call site.
if (self->state.stage)
  shutdown_stream_stage(*self->state.stage);
```

I don't have a strong preference, but if we have a former than there should be a comment above the function that it is a no-op if the stage is a nullptr.